File: frontend/lib/domain/entities/signal.dart
enum SignalStatus { active, pending, closed, draft }

extension SignalStatusLabel on SignalStatus {
  String get apiValue => name.toUpperCase();

  String get label => switch (this) {
    SignalStatus.active => 'Active',
    SignalStatus.pending => 'Pending',
    SignalStatus.closed => 'Closed',
    SignalStatus.draft => 'Draft',
  };

  static SignalStatus parse(String value) {
    return switch (value.toUpperCase()) {
      'PENDING' => SignalStatus.pending,
      'CLOSED' => SignalStatus.closed,
      'DRAFT' => SignalStatus.draft,
      _ => SignalStatus.active,
    };
  }
}

enum PositionSide { long, short }

extension PositionSideLabel on PositionSide {
  String get apiValue => name.toUpperCase();

  String get label => name.toUpperCase();

  static PositionSide parse(String value) {
    return value.toUpperCase() == 'SHORT'
        ? PositionSide.short
        : PositionSide.long;
  }
}

enum SignalGrade { strongBuy, buy, watch, noTrade }

extension SignalGradeLabel on SignalGrade {
  String get label => switch (this) {
    SignalGrade.strongBuy => 'STRONG BUY',
    SignalGrade.buy => 'BUY',
    SignalGrade.watch => 'WATCH',
    SignalGrade.noTrade => 'NO TRADE',
  };

  static SignalGrade? parse(String? value) {
    if (value == null) return null;
    return switch (value.toUpperCase()) {
      'STRONG_BUY' => SignalGrade.strongBuy,
      'BUY' => SignalGrade.buy,
      'WATCH' => SignalGrade.watch,
      'NO_TRADE' => SignalGrade.noTrade,
      _ => null,
    };
  }
}

enum EntryType { preBreakout, breakout, breakoutRetest }

extension EntryTypeLabel on EntryType {
  String get label => switch (this) {
    EntryType.preBreakout => 'Pre-Breakout',
    EntryType.breakout => 'Breakout',
    EntryType.breakoutRetest => 'Breakout Retest',
  };

  static EntryType? parse(String? value) {
    if (value == null) return null;
    return switch (value.toUpperCase()) {
      'BREAKOUT' => EntryType.breakout,
      'BREAKOUT_RETEST' => EntryType.breakoutRetest,
      _ => EntryType.preBreakout,
    };
  }
}

enum MarketRegime { bullish, neutral, bearish }

extension MarketRegimeLabel on MarketRegime {
  String get label => switch (this) {
    MarketRegime.bullish => 'Bullish',
    MarketRegime.neutral => 'Neutral',
    MarketRegime.bearish => 'Bearish',
  };

  static MarketRegime? parse(String? value) {
    if (value == null) return null;
    return switch (value.toUpperCase()) {
      'BULLISH' => MarketRegime.bullish,
      'BEARISH' => MarketRegime.bearish,
      _ => MarketRegime.neutral,
    };
  }
}

class Signal {
  const Signal({
    required this.id,
    required this.symbol,
    required this.status,
    required this.side,
    this.confidence,
    this.entryPrice,
    this.targetPrice,
    this.stopLoss,
    this.strategy,
    this.strategyWinRate,
    this.suggestedRiskPercent,
    this.closedAt,
    this.technicalSummary,
    this.disclaimer,
    required this.createdAt,
    // Spot Morning Plan fields
    this.score,
    this.signalGrade,
    this.entryType,
    this.marketRegime,
    this.targetPrice2,
    this.targetPrice3,
    this.riskReward,
  });

  final String id;
  final String symbol;
  final SignalStatus status;
  final PositionSide side;
  final int? confidence;
  final double? entryPrice;
  final double? targetPrice;
  final double? stopLoss;
  final String? strategy;
  final double? strategyWinRate;
  final double? suggestedRiskPercent;
  final DateTime? closedAt;
  final String? technicalSummary;
  final String? disclaimer;
  final DateTime createdAt;

  // Spot Morning Plan
  final int? score;
  final SignalGrade? signalGrade;
  final EntryType? entryType;
  final MarketRegime? marketRegime;
  final double? targetPrice2;
  final double? targetPrice3;
  final double? riskReward;
}

File: backend/src/main/java/com/shyblack/cryptosignals/entity/Signal.java
package com.shyblack.cryptosignals.entity;

import com.shyblack.cryptosignals.entity.enums.EntryType;
import com.shyblack.cryptosignals.entity.enums.MarketRegime;
import com.shyblack.cryptosignals.entity.enums.PositionSide;
import com.shyblack.cryptosignals.entity.enums.SignalGrade;
import com.shyblack.cryptosignals.entity.enums.SignalStatus;
import com.shyblack.cryptosignals.entity.enums.TradingMode;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.Instant;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@Entity
@Table(name = "signals")
public class Signal extends BaseEntity {

	@Column(nullable = false)
	private String symbol;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	private SignalStatus status = SignalStatus.DRAFT;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	private PositionSide side;

	@Column(nullable = false)
	private Integer confidence;

	@Column(nullable = false, precision = 19, scale = 8)
	private BigDecimal entryPrice;

	@Column(nullable = false, precision = 19, scale = 8)
	private BigDecimal targetPrice;

	@Column(nullable = false, precision = 19, scale = 8)
	private BigDecimal stopLoss;

	private String strategy;

	@Column(precision = 7, scale = 4)
	private BigDecimal strategyWinRate;

	@Column(nullable = false, precision = 5, scale = 2)
	private BigDecimal suggestedRiskPercent = new BigDecimal("2.00");

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "created_by_user_id")
	private User createdBy;

	private Instant closedAt;

	@Column(length = 2000)
	private String technicalSummary;

	@Column(length = 1000)
	private String disclaimer;

	// --- Spot Morning Plan fields ---

	@Column
	private Integer score;

	@Enumerated(EnumType.STRING)
	@Column
	private SignalGrade signalGrade;

	@Enumerated(EnumType.STRING)
	@Column
	private EntryType entryType;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	private TradingMode tradingMode = TradingMode.SPOT;

	@Enumerated(EnumType.STRING)
	@Column
	private MarketRegime marketRegime;

	@Column(precision = 19, scale = 8)
	private BigDecimal targetPrice2;

	@Column(precision = 19, scale = 8)
	private BigDecimal targetPrice3;

	@Column(precision = 10, scale = 4)
	private BigDecimal riskReward;
}
